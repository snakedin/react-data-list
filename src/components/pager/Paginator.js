import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withOptions } from "../../options/withOptions";

class Paginator extends Component {

    static propTypes = {
        page: PropTypes.number,
        pageSize: PropTypes.number,
        totalCount: PropTypes.number,
        onParamChanged: PropTypes.func
    };

    /**
     * Determine pages set to show
     */
    getPageSet(totalPages, pageShowMax, currentPage) {

        let pageFrom = 1;
        let pageTo = totalPages;
        let pageSet = [];

        if (totalPages > pageShowMax) {

            let halfPage = Math.floor(pageShowMax / 2);

            if (currentPage > halfPage && currentPage <= (totalPages - halfPage)) {
                pageFrom = currentPage - halfPage;
                pageTo = pageFrom + pageShowMax - 1;
            } else if (currentPage <= halfPage) {
                pageFrom = 1;
                pageTo = pageFrom + pageShowMax - 1;
            } else if (currentPage >= (totalPages - halfPage)) {
                pageTo = totalPages;
                pageFrom = pageTo - pageShowMax + 1;
            }
        }

        for (let i = pageFrom;  i <= pageTo; i++) {
            pageSet.push(i);
        }

        return pageSet;
    }

    /**
     * Render previous page button
     */
    getPreviousButton() {

        const { page, onParamChanged } = this.props;
        const label = (
            <span aria-hidden="true">&laquo;</span>
        );

        if (page <= 1) {
            return (
                <li className="page-item disabled" key="prev">
                    <button className="page-link" tabIndex="-1">{label}</button>
                </li>
            )
        } else {
            return (
                <li className="page-item" key="prev">
                    <button className="page-link" aria-label="Previous" onClick={ () => { onParamChanged('page', page-1) } }>{label}</button>
                </li>
            );
        }
    };

    /**
     * Render next page button
     */
    getNextButton(totalPages) {

        const { page, onParamChanged } = this.props;
        const label = (
            <span aria-hidden="true">&raquo;</span>
        );

        if (page === totalPages) {
            return (
                <li className="page-item disabled" key="next">
                    <button className="page-link" tabIndex="-1">{label}</button>
                </li>
            )
        } else {
            return (
                <li className="page-item" key="next">
                    <button className="page-link" aria-label="Next" onClick={ () => { onParamChanged('page', page+1) } }>{label}</button>
                </li>
            );
        }
    };

    /**
     * Render element
     */
    render() {

        const {
            totalCount, page, pageSize, onParamChanged,
            options: { classNamePager, pageShowMax, showPageNext, showPagePrevious }
        } = this.props;

        const totalPages = Math.ceil(totalCount / pageSize);
        let pager = [];

        if (totalPages > 1) {

            // Previous page
            if (showPagePrevious) {
                pager.push(this.getPreviousButton());
            }

            // Pager
            const pageSet = this.getPageSet(totalPages, pageShowMax, page);

            pageSet.forEach((value) => {
                let className = 'page-item';

                if (page === value)
                    className += ' active';

                pager.push((
                    <li key={value} className={className}>
                        <button className="page-link" onClick={ () => { onParamChanged('page', value) } }>
                            { value }
                        </button>
                    </li>
                ));
            });

            // Next page
            if (showPageNext) {
                pager.push(this.getNextButton(totalPages));
            }
        }

        return (
            <ul className={ classNamePager }>
                {
                    pager.map((element) => {
                        return element;
                    })
                }
            </ul>
        );
    }

}

export default withOptions(Paginator);