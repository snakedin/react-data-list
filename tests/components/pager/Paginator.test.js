import React from "react";
import { render } from "@testing-library/react";
import "../../mocks/withOptionsMock";
import Paginator from "../../../src/components/pager/Paginator";

it("pager/Paginator render", () => {
  const { asFragment, rerender } = render(<Paginator page={1} pageSize={10} totalCount={100}/>);
  const firstPage = asFragment();

  rerender(<Paginator page={5} pageSize={10} totalCount={100}/>);
  const middlePage = asFragment();

  rerender(<Paginator page={10} pageSize={10} totalCount={100}/>);
  const lastPage = asFragment();

  expect(firstPage).toMatchSnapshot();
  expect(firstPage).toMatchDiffSnapshot(middlePage);
  expect(middlePage).toMatchDiffSnapshot(lastPage);
});
