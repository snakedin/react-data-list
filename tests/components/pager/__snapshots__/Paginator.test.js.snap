// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pager/Paginator render 1`] = `
<DocumentFragment>
  <ul
    class="pagination mb-0 justify-content-center justify-content-lg-start"
  >
    <li
      class="page-item disabled"
    >
      <button
        class="page-link"
        tabindex="-1"
      >
        <span
          aria-hidden="true"
        >
          «
        </span>
      </button>
    </li>
    <li
      class="page-item active"
    >
      <button
        class="page-link"
      >
        1
      </button>
    </li>
    <li
      class="page-item"
    >
      <button
        class="page-link"
      >
        2
      </button>
    </li>
    <li
      class="page-item"
    >
      <button
        class="page-link"
      >
        3
      </button>
    </li>
    <li
      class="page-item"
    >
      <button
        class="page-link"
      >
        4
      </button>
    </li>
    <li
      class="page-item"
    >
      <button
        class="page-link"
      >
        5
      </button>
    </li>
    <li
      class="page-item"
    >
      <button
        class="page-link"
      >
        6
      </button>
    </li>
    <li
      class="page-item"
    >
      <button
        class="page-link"
      >
        7
      </button>
    </li>
    <li
      class="page-item"
    >
      <button
        class="page-link"
      >
        8
      </button>
    </li>
    <li
      class="page-item"
    >
      <button
        class="page-link"
      >
        9
      </button>
    </li>
    <li
      class="page-item"
    >
      <button
        class="page-link"
      >
        10
      </button>
    </li>
    <li
      class="page-item"
    >
      <button
        aria-label="Next"
        class="page-link"
      >
        <span
          aria-hidden="true"
        >
          »
        </span>
      </button>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`pager/Paginator render 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,25 +1,25 @@
  <DocumentFragment>
    <ul
      class=\\"pagination mb-0 justify-content-center justify-content-lg-start\\"
    >
      <li
-       class=\\"page-item disabled\\"
+       class=\\"page-item\\"
      >
        <button
+         aria-label=\\"Previous\\"
          class=\\"page-link\\"
-         tabindex=\\"-1\\"
        >
          <span
            aria-hidden=\\"true\\"
          >
            «
          </span>
        </button>
      </li>
      <li
-       class=\\"page-item active\\"
+       class=\\"page-item\\"
      >
        <button
          class=\\"page-link\\"
        >
          1
@@ -51,11 +51,11 @@
        >
          4
        </button>
      </li>
      <li
-       class=\\"page-item\\"
+       class=\\"page-item active\\"
      >
        <button
          class=\\"page-link\\"
        >
          5"
`;

exports[`pager/Paginator render 3`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -51,11 +51,11 @@
        >
          4
        </button>
      </li>
      <li
-       class=\\"page-item active\\"
+       class=\\"page-item\\"
      >
        <button
          class=\\"page-link\\"
        >
          5
@@ -96,24 +96,24 @@
        >
          9
        </button>
      </li>
      <li
-       class=\\"page-item\\"
+       class=\\"page-item active\\"
      >
        <button
          class=\\"page-link\\"
        >
          10
        </button>
      </li>
      <li
-       class=\\"page-item\\"
+       class=\\"page-item disabled\\"
      >
        <button
-         aria-label=\\"Next\\"
          class=\\"page-link\\"
+         tabindex=\\"-1\\"
        >
          <span
            aria-hidden=\\"true\\"
          >
            »"
`;
